core/src/macros.rs

```diff
@@ -210,39 +210,54 @@ macro_rules! poly_fn {
     => { poly_fn!(
        p~ [$($tparams, )*] |$arg: $arg_typ| -> $ret_typ {$body}, ~p  f~ ~f $($rest)*
     )};
+    // encountered first func w/ type params, trailing comma on tparams
+    ([$($tparams: tt, )*] |$arg: ident : $arg_typ: ty| -> $ret_typ: ty { $body: expr }, $($rest: tt)*)
```

Comment:
`$($tparams: tt, )*` seems fishy to me, does this work for something like `T: Clone`?

Reply:
No, sadly it doesn't :(

Reply:
Thinking about this a slight bit more, the macro may also need support for `where` bounds, because not all bounds can be put in the type parameter list.  Unfortunately, I also recall that `where` bounds were one of *the most painful* design considerations in [`unbox_macro`](https://github.com/ExpHP/unbox-macro), because, much like type parameter lists, they're impossible to parse with `macro_rules!`.

Reply:
Yeah, .I'm getting flashbacks of trying to implement LabelledGeneric + Generic derivations with macros before custom derives were stable.

Until we have proc macros (not stable yet I think?), maybe we should just include this macro, but set limits on what it can do, beyond which we just tell users to go the long route (write the derives by hand)?

Reply:
I agree with setting limits.  There's only so much that a `macro_rules` macro can reasonably promise without turning into a maintenance black hole